src/cromshell/cost/command.py

```diff
@@ -0,0 +1,302 @@
+        raise ValueError(
+            f"Could not find workflow id {resolved_workflow_id} in submission file."
+        )
+
+    # check that bq setting in json config
+    check_cost_table_is_configured(config_options=config.cromshell_config_options)
+
+    # Get time workflow finished using metadata command (error if not finished)
+    LOGGER.info("Retrieving workflow metadata")
+    workflow_metadata = get_metadata(config=config)
+    start_time, end_time = get_submission_start_end_time(workflow_metadata=workflow_metadata)
+
+    LOGGER.info("Checking workflow completed and finished past 24hrs")
+    checks_before_query(end_time=end_time, workflow_id=resolved_workflow_id)
+
+    LOGGER.info("Querying BQ")
+    # Create start and end time for query, plus/minus a day from start and finish
+    query_start_time = str(datetime.strptime(start_time, "%Y-%m-%dT%H:%M:%S.%f%z") - timedelta(days=1))
+    query_end_time = str(datetime.strptime(end_time, "%Y-%m-%dT%H:%M:%S.%f%z") + timedelta(days=1))
+
+    query_results = query_bigquery(
+        workflow_id=resolved_workflow_id,
+        bq_database=config.cromshell_config_options["bq_cost_table"],
+        start_date=query_start_time,
+        end_date=query_end_time,
+        detailed=detailed
+    )
+
+    temp_query_result_csv = NamedTemporaryFile()
+    write_bq_results_to_temp_csv(
+        tempfile=temp_query_result_csv, results=query_results, detailed=detailed
+    )
+
+    temp_query_result_csv_rounded = NamedTemporaryFile()
+    round_cost_values(
+        query_result_csv=temp_query_result_csv,
+        query_result_csv_rounded=temp_query_result_csv_rounded,
+        detailed=detailed
+    )
+
+    with open(temp_query_result_csv_rounded.name, 'r') as f:
+        print(f.read())
+        print("Costs rounded to nearest cent (approximately).")
+
+
+def print_query_results(query_results, detailed: bool) -> None:
+    # performing a step like getting the first row from the table would require to convert
+    # the table to datafram or CSV. Instead, this loop will get the first row and break.
+
+    if detailed:
+        for row in query_results:
+            print("{}\t{}\t{}\t{}".format(row.value, row.description, row.task_name,
+                                              row.cost))
+
+    else:
+        for row in query_results:
+            total_cost = round(float(row.cost), 2)
+            print(f"${total_cost}")
+            print("Costs rounded to nearest cent (approximately).")
+
+
+def query_bigquery(
+    workflow_id: str,
+    bq_database: str,
+    start_date: str,
+    end_date: str,
+    detailed: bool
+):
+    """
+        Query bigqueury for cost
+
+    :param workflow_id:
+    :param bq_database:
+    :param start_date:
+    :param end_date:
+    :return:
+    """
+    from google.cloud import bigquery
+    client = bigquery.Client()
+
+    # Todo: remove LIMIT in query
+    if detailed:
+        query_job = client.query(
```

Comment:
is there a structured query builder AP so you're not doing direct string interpolation?  It's *probably* not an issue but queries built like this are insanely vulnerable to sql injection type attacks.  Unless of course python is clever enough to use a custom interpolation engine here which performs escaping.  I have heard about for some languages but I'm not sure which.

Reply:
Good call! There's some stuff about it [here](https://cloud.google.com/bigquery/docs/parameterized-queries) that can be implemented. I was able to parameterize all user input except table name, which the article expertly says it can't do.